staging/src/k8s.io/apiserver/pkg/admission/metrics.go

```diff
@@ -0,0 +1,161 @@
+		"Admission controller %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	// External admission webhook metrics. Each webhook is identified by name.
+	externalWebhookMetrics = newAdmissionMetrics("external_webhook_",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource"},
+		"External admission webhook %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+)
+
+func init() {
+	stepMetrics.mustRegister()
+	controllerMetrics.mustRegister()
+	externalWebhookMetrics.mustRegister()
+}
+
+// namedHandler requires each admission.Interface be named, primarly for metrics tracking purposes.
+type NamedHandler interface {
+	Interface
+	GetName() string
+}
+
+// ObserveAdmissionStep records admission related metrics for a admission step, identified by step type.
+func ObserveAdmissionStep(elapsed time.Duration, rejected bool, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	stepMetrics.observe(elapsed, rejected, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), stepType)
+}
+
+// ObserveAdmissionController records admission related metrics for a build-in admission controller, identified by it's plugin handler name.
+func ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes) {
+	t := typeToLabel(handler)
+	gvr := attr.GetResource()
+	controllerMetrics.observe(elapsed, rejected, handler.GetName(), t, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource())
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.ExternalAdmissionHook, attr Attributes) {
+	t := "validating" // TODO: pass in type (validating|mutating) once mutating webhook functionality has been implemented
+	gvr := attr.GetResource()
+	externalWebhookMetrics.observe(elapsed, rejected, hook.Name, t, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource())
+}
+
+func typeToLabel(i Interface) string {
+	switch i.(type) {
+	case ValidationInterface:
+		return "validating"
+	case MutationInterface:
+		return "mutating"
+	default:
+		return "UNRECOGNIZED_ADMISSION_TYPE"
+	}
+}
+
+type admissionMetrics struct {
+	total            *prometheus.CounterVec
+	rejectedTotal    *prometheus.CounterVec
+	latencies        *prometheus.HistogramVec
+	latenciesSummary *prometheus.SummaryVec
+}
+
+func (m *admissionMetrics) mustRegister() {
+	prometheus.MustRegister(m.total)
+	prometheus.MustRegister(m.rejectedTotal)
+	prometheus.MustRegister(m.latencies)
+	prometheus.MustRegister(m.latenciesSummary)
+}
+
+func (m *admissionMetrics) observe(elapsed time.Duration, rejected bool, labels ...string) {
```

Comment:
is the order of the labels necessary?

Reply:
Yeah, the labels are positional. They're defined using a `[]string` in calls to functions like `NewCounterVec` and are assigned values via the varadic `WithLabelValues` function. It's not type safe, which is why I've hidden this `observe` method behind strongly typed `Observe...()` functions.